mitosheet/mitosheet/mito_dash/v1/spreadsheet.py

```diff
@@ -0,0 +1,202 @@
+        
+            
+    def process_single_message(self):
+
+        # If we are already processing messages -- then wait until it is
+        if self.processing_messages:
+            while self.processing_messages:
+                time.sleep(0.1)
+            
+        # Otherwise, set the processing flag to true
+        self.processing_messages = True
+
+        # Process all the messages in the queue
+        try:
+            if not self.unprocessed_messages.empty():
+                value = self.unprocessed_messages.get()
+                self.mito_backend.receive_message(value)
+        except:
+            # Make sure we always set the processing flag to false
+            pass
+            
+        # Set the processing flag to false
+        self.processing_messages = False
+        
+        
+    def get_all_json(self):
+        return json.dumps({
+            **self.mito_backend.get_shared_state_variables(),
+            'responses_json': json.dumps(self.responses),
+            'key': self.mito_frontend_key
+        })
+    
+    def get_result(self):
+        return SpreadsheetResult(
+            dfs=self.mito_backend.steps_manager.dfs,
+            code=self.mito_backend.steps_manager.code(),
+        )
+    
+
+def spreadsheet_callback(
+    *_args,
+    input_id=None
+):
+    """
+    Provides a server-side callback relating the values of one or more `Output` items 
+    to a single spreadsheet with the given id.
+
+    Example usage:
+    ```
+    from mitosheet.mito_dash.v1 import Spreadsheet, spreadsheet_callback
+    from dash import Dash, html, Output
+
+    import pandas as pd
+
+    df = pd.DataFrame({'A': [1, 2, 3]})
+
+    app = Dash(__name__)
+
+    app.layout = html.Div([
+        Spreadsheet(df, id='mito-dash-wrapper'),
+        html.Div(id='output'),    
+    ])
+
+    @spreadsheet_callback(
+        Output('output', 'children'),
+        input_id='mito-dash-wrapper',
+    )
+    def update_output(spreadsheet_result):
+        dfs = spreadsheet_result.dfs()
+        return f'Output: {str(dfs)}'
+
+    
+    if __name__ == '__main__':
+        app.run_server(debug=True)
+    ```
+
+    """
+    if input_id is None:
+        raise ValueError('input_id must be provided')
+    
+    # Find the Spreadsheet component with the given id - searching all objects in this Python runtime
+    callback_component = None
+    components = [
+        obj for obj in gc.get_objects()
+        if isinstance(obj, Spreadsheet) and getattr(obj, 'mito_id', None) == input_id
+    ]
+
+    if len(components) > 0:
+        callback_component = components[0]
+
+    if callback_component is None:
+        raise ValueError(f'Could not find Spreadsheet with id {input_id}')
+    
+    # Create a callback that will send the result to the Spreadsheet component
+    # with the given id
+    def callback_decorator(func):
+        @callback(
+            *_args,
+            Input(callback_component.mito_id, 'all_json'),
```

Comment:
What is all_json?

Reply:
It's just a collection of all the props we send to the mitosheet. We package them so we can do a single update at once, and because it's easy to split them apart. 

There's a type on the frontend if you want to see what's in it.